proposals/shared-everything-threads/Overview.md

```diff
@@ -702,6 +702,35 @@ models. For more discussion: [How should toolchains apply `shared`?][toolchain-s
 
 [toolchain-shared-discussion]: https://github.com/abrown/thread-spawn/discussions/5
 
+### How will GCs and engines need to change?
+
+GCs that historically assume that there is only a single mutator thread will need to handle heaps
+with multiple mutators running in parallel. Since sharedness is part of static types, engines can
+know at allocation or access time when they need to do something different for shared objects, so it
+is possible for them to add a separate concept of shared heap on top of their existing
+single-threaded heaps. This might be a reasonable intermediate architecture, even for engines that
+plan on eventually switching to having shared heaps as their only primitive.
+
+References from shared objects to unshared are normally disallowed, but we could choose to support
+them in limited cases to make TLS or thread-bound data more ergonomic. If we do, there are varying
+guarantees implementations could make about how the GC treats these references. (As usual, the spec
+will say nothing about how garbage is collected.)
+
+The weakest behavior would be that shared-to-unshared references never suffice to keep the unshared
+object alive, i.e. they are only allowed as weak references. Users would have to root unshared
+objects for the duration of their intended lifetimes, so this is no more expressive than simply
+disallowing shared-to-unshared references.
+
+The strongest behavior would be that shared-to-unshared references behave exactly like normal
+references with respect to garbage collection. To collect cross-heap cycles, the garbage collector
+would need visibility over the heaps from all threads.
+
+If shared objects are allowed as keys in `FinalizationRegistry`, then the weak behavior would be
+equivalent in expressiveness to the strong behavior, except that it would not be able to collect
+cross-heap cycles. The `FinalizationRegistry` would be able to automatically manage the lifetimes of
+rooted unshared objects as long as they do not cyclically keep the shared objects that refer to them
+alive.
```

Comment:
I'm still not entirely happy with this last section about `FinalizationRegistry`.

What's in the paragraph is true, provided "allowed as keys in `FinalizationRegistry`" means "allowed as keys in `FinalizationRegistry` and either disallowed as keys in `WeakMap`s, or cycles caused by `WeakMap`s also leak". For the former, I (and others in TC39, I suspect) wouldn't be okay with allowing shared objects in `FinalizationRegistry` and `WeakRef` and *not* also as keys in `WeakMap`, because it would look like surprising and arbitrary non-compositionality of features.

All that's to say I'd caveat it with, while strictly possible, this particular outcome where shared objects are usable in _some_ weak APIs but not in others is unlikely to happen.

Reply:
I've rewritten this paragraph to be very explicit that this point is only about relating the strong behavior without cross-heap cycle collection to the weak behavior and that we would not ship with support for only `FinalizationRegistry`.